File: tool/coding/backend/regex_search.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""regex_search - 正则搜索文件内容"""
import os
import re
from tool.coding.backend.base import ToolContext

TOOL_NAME = 'regex_search'

_SKIP_DIRS = {'.git', 'node_modules', '__pycache__', '.venv', 'venv'}
_TEXT_EXTS = {'.py', '.js', '.ts', '.tsx', '.jsx', '.java', '.c', '.cpp', '.h', '.hpp',
              '.go', '.rs', '.rb', '.php', '.css', '.scss', '.html', '.htm', '.xml',
              '.json', '.yaml', '.yml', '.toml', '.ini', '.cfg', '.md', '.txt', '.sh',
              '.bat', '.ps1', '.sql', '.vue', '.svelte'}


def _regex_search_file(filepath, pattern, case_insensitive, context_lines, show_groups, max_results):
    matches = []
    try:
        with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
            lines = f.readlines()
    except Exception:
        return matches

    for i, line in enumerate(lines):
        if len(matches) >= max_results:
            break
        m = pattern.search(line)
        if m:
            ctx_start = max(0, i - context_lines)
            ctx_end = min(len(lines), i + context_lines + 1)
            entry = {
                'line': i + 1,
                'match': m.group(0).rstrip('\n\r'),
                'context': [l.rstrip('\n\r') for l in lines[ctx_start:ctx_end]]
            }
            if show_groups and m.groups():
                entry['groups'] = list(m.groups())
            matches.append(entry)
    return matches


def handle(body, ctx):
    try:
        pattern_str = body.get('pattern', '')
        if not pattern_str:
            ctx.send_error('pattern is required')
            return

        paths = body.get('paths')
        if not paths:
            p = body.get('path', '')
            paths = [p] if p else []
        if not paths:
            ctx.send_error('No path specified')
            return

        case_insensitive = body.get('case_insensitive', False)
        max_results = int(body.get('max_results', 50))
        context_lines = int(body.get('context_lines', 2))
        show_groups = body.get('show_groups', True)
        file_type = body.get('file_type', None)

        flags = re.IGNORECASE if case_insensitive else 0
        try:
            pattern = re.compile(pattern_str, flags)
        except re.error as e:
            ctx.send_error('Invalid regex: ' + str(e))
            return

        all_results = []
        total_matches = 0

        for search_path in paths:
            if total_matches >= max_results:
                break
            if os.path.isfile(search_path):
                if file_type and not search_path.endswith(file_type):
                    continue
                matches = _regex_search_file(search_path, pattern, case_insensitive, context_lines, show_groups, max_results - total_matches)
                if matches:
                    all_results.append({'file': search_path, 'matches': matches})
                    total_matches += len(matches)
            elif os.path.isdir(search_path):
                for dirpath, dirnames, filenames in os.walk(search_path):
                    dirnames[:] = [d for d in dirnames if d not in _SKIP_DIRS and not d.startswith('.')]
                    for fn in sorted(filenames):
                        if total_matches >= max_results:
                            break
                        if file_type and not fn.endswith(file_type):
                            continue
                        if not file_type:
                            ext = os.path.splitext(fn)[1].lower()
                            if ext not in _TEXT_EXTS:
                                continue
                        fp = os.path.join(dirpath, fn)
                        matches = _regex_search_file(fp, pattern, case_insensitive, context_lines, show_groups, max_results - total_matches)
                        if matches:
                            all_results.append({'file': fp, 'matches': matches})
                            total_matches += len(matches)

        ctx.send_json({'ok': True, 'results': all_results, 'total_matches': total_matches})
    except Exception as e:
        ctx.send_error(str(e))

File: tools/coding/frontend/regex_search.js
// ========== regex_search.js - 工具定义 ==========
// 朱峰社区智能体无限 5.0.0 - 独立工具文件
// 自动生成自 4.1.8 tools.js（split_tools.js V3）
// 执行逻辑: tools/executor.js 中对应 case 分支

window.Tools = window.Tools || { allTools: {} };
if (!window.Tools.allTools['regex_search']) {
    window.Tools.allTools['regex_search'] = {
    regex_search: {
        type: 'function',
        function: {
            name: 'regex_search',
            description: '正则表达式搜索工具（始终正则模式）。支持文件/文件夹、单路径/多路径。输出匹配行、捕获组、上下文。比 search_in_files+regex=true 更强大：自动提取捕获组、高亮匹配位置。',
            parameters: {
                type: 'object',
                properties: {
                    pattern: {
                        type: 'string',
                        description: '正则表达式（必填）。如 \\bdef\\s+(\\w+) 匹配Python函数定义'
                    },
                    path: {
                        type: 'string',
                        description: '（单个）文件或目录路径'
                    },
                    paths: {
                        type: 'array',
                        items: { type: 'string' },
                        description: '（多个）文件或目录路径数组'
                    },
                    case_insensitive: {
                        type: 'boolean',
                        description: '是否不区分大小写（默认 false）'
                    },
                    max_results: {
                        type: 'integer',
                        description: '最大返回匹配数（默认 50）'
                    },
                    context_lines: {
                        type: 'integer',
                        description: '每个匹配显示几行上下文（默认 2）'
                    },
                    file_type: {
                        type: 'string',
                        description: '按扩展名过滤搜索文件，如 .py 或 .js'
                    },
                    show_groups: {
                        type: 'boolean',
                        description: '是否显示捕获组详情（默认 true）'
                    }
                },
                required: ['pattern']
            }
        }
    },

    // ===== 工单清单工具（多轮规划，批量执行） =====
    };
}
